dali/python/nvidia/dali/plugin/mxnet.py

```diff
@@ -423,3 +459,224 @@ def __init__(self,
+
+    fill_last_batch = True, last_batch_padded = True   -> last batch = ``[7, 7]``, next iteration will return ``[1, 2]``
+
+    fill_last_batch = True, last_batch_padded = False  -> last batch = ``[7, 1]``, next iteration will return ``[2, 3]``
+
+    """
+    def __init__(self,
+                 pipelines,
+                 size,
+                 output_types=None,
+                 auto_reset=False,
+                 fill_last_batch=True,
+                 last_batch_padded=False):
+        super(DALIGluonIterator, self).__init__(
+            pipelines,
+            size,
+            fill_last_batch,
+            last_batch_padded,
+            auto_reset)
+
+        # Build all pipelines
+        for p in self._pipes:
+            with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
+                p.build()
+        self._data_batches = [None for i in range(self._num_gpus)]
+        self._counter = 0
+        self._output_tags = {DALIGluonIterator.DENSE_TAG, DALIGluonIterator.SPARSE_TAG}
+        assert output_types is None or set(output_types) <= self._output_tags, \
+            "Only DENSE_TAG and SPARSE_TAG are allowed"
+
+        self._outputs_types = output_types
+
+        # We need data about the batches (like shape information),
+        # so we need to run a single batch as part of setup to get that info
+        for p in self._pipes:
+            with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
+                p.schedule_run()
+        self._first_batch = None
+        self._first_batch = self.next()
+
+
+    def __next__(self):
+        if self._first_batch is not None:
+            batch = self._first_batch
+            self._first_batch = None
+            return batch
+        if self._counter >= self._size and self._size > 0:
+            if self._auto_reset:
+                self.reset()
+            raise StopIteration
+        # Gather outputs
+        dali_outputs = []
+        for p in self._pipes:
+            with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
+                dali_outputs.append(p.share_outputs())
+        for i in range(self._num_gpus):
+            output_elements = []
+            shapes = []
+            for j, out in enumerate(dali_outputs[i]):
+                if self._outputs_types is None or self._outputs_types[j] == DALIGluonIterator.DENSE_TAG:
+                    output_elements.append(out.as_tensor())
+                    shapes.append(output_elements[-1].shape())
+                else:
+                    output_elements.append([out[sample_idx] for sample_idx in range(self.batch_size)])
+                    s = [t.shape() for t in output_elements[-1]]
+                    shapes.append(s)
+
+            if self._data_batches[i] is None:
+                self._data_batches[i] = self._create_data_batch(output_elements, shapes, self._pipes[i].device_id)
```

Comment:
In the GenericIterator we use double buffering for whatever reason. Just want to make sure if having one copy of the buffers that we fill and hand out to MXNet is safe?

Reply:
We are not since the multiple buffering size is 1 😉 It used to be the case but not anymore